tripy/tripy/frontend/utils.py

```diff
@@ -193,6 +194,20 @@ def find_sync_target_dtype(arg_name):
                     return None
 
                 def convert_nontensor_arg(arg, list_index=None):
+                    def is_valid_sequence(seq_arg: Sequence) -> bool:
+                        if len(seq_arg) == 0:
+                            return True
+                        if isinstance(seq_arg[0], Sequence):
+                            return is_valid_sequence(seq_arg[0])
+                        return isinstance(seq_arg[0], (int, float))
+
+                    if not isinstance(arg, (int, float)) and not (isinstance(arg, Sequence) and is_valid_sequence(arg)):
+                        raise_error(
```

Comment:
I think a better way to do this might be to just `return arg`. Then the function registry's type checking will take over and generate potentially nicer error messages (we might need to update the type annotations to `Union[Tensor, numbers.Number]` and then also make the registry handle `Union` types (basically, in `matches_type` we would just do `any(matches_type(...))` for the contained types

Reply:
We would need to update the registry to have better error messages for a case like this. Here is a failure I get if I try this approach:

```python
import tripy as tp
import numpy as np

a = tp.Tensor([1, 2, 3])
b = np.array([1, 2, 3])
print(a + b)
```

Error:
```
  File "/tripy/tripy/function_registry.py", line 281, in wrapper
    return self.find_overload(key, args, kwargs)(*args, **kwargs)
  File "/tripy/tripy/function_registry.py", line 185, in __call__
    return self.func(*args, **kwargs)
  File "/tripy/tripy/function_registry.py", line 281, in wrapper
    return self.find_overload(key, args, kwargs)(*args, **kwargs)
  File "/tripy/tripy/function_registry.py", line 185, in __call__
    return self.func(*args, **kwargs)
  File "/tripy/tripy/frontend/utils.py", line 273, in wrapper
    return func(*new_args, **new_kwargs)
  File "/tripy/tripy/frontend/trace/ops/binary_elementwise.py", line 208, in __add__
    return BinaryElementwise.build([self, other], BinaryElementwise.Kind.SUM)
  File "/tripy/tripy/frontend/trace/ops/base.py", line 77, in build
    inp_trace_tensors = [inp.trace_tensor for inp in inputs]
  File "/tripy/tripy/frontend/trace/ops/base.py", line 77, in <listcomp>
    inp_trace_tensors = [inp.trace_tensor for inp in inputs]
AttributeError: 'numpy.ndarray' object has no attribute 'trace_tensor'
```

Reply:
I think that's only for the case of properties like `__add__` because this error message seems sensible:

```python
import tripy as tp

b = np.array([1, 2, 3])
print(tp.reshape(b, (1, 3)))
```

Error:
```
Could not find an implementation for function: 'reshape'.
    Note: Argument types were: [ndarray, tuple].
    Candidate overloads were:

    --> /tripy/tripy/frontend/trace/ops/reshape.py:75 in reshape()
          |
       75 | def reshape(input: "tripy.Tensor", shape: Union["tripy.Shape", Sequence[Union[int, "tripy.Tensor"]]]) -> "tripy.Tensor":
       76 |     ...
          | 

    Not a valid overload because: For parameter: 'input', expected an instance of type: 'Tensor' but got argument of type: 'ndarray'.
```

Reply:
Correctly checking convertible tensor literals would require a bit of extra machinery, since you could have a sequence of any depth. This is possible to handle, thankfully: https://stackoverflow.com/questions/53845024/defining-a-recursive-type-hint-in-python

Reply:
I added in support for `Union` and `Sequence` in the checker (the generated messages are very clunky, but we could make further changes) and put in a WAR for tensor literals (general support for recursive annotations is probably a rabbithole we don't want to go down)

Reply:
Do you have an example of the generated error messages?

Reply:
I actually think it's more clear to just show the full annotation:
```
Not a valid overload because: For parameter: 'n', expected an instance of type: 'Sequence[Union[int, float]]' but got: <...>.
```

Reply:
I had the more detailed report because it will oftentimes give something less detailed like:

```
For parameter 'n', expected an instance of type typing.Sequence[typing.Union[int, float]] but got list
```

That's just how `type` works. However, this suggests that doing a little bit of extra work in the sequence case could pay off, so I'll try that instead 🙂

Reply:
Improving the pretty-printing in the type matcher got rid of the need for weird recursive messages like these

Reply:
I've used the approach from #189 (`__forward_arg__`) to handle the recursive case directly, so no need for a WAR now 🙂The signature is a little clunky because it needed to be inside a class to be exported and documented (`tripy.TensorLiteral.sig`).